editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-parent-lookup.ts

```diff
@@ -0,0 +1,482 @@
+import { getStoryboardElementPath } from '../../../../../core/model/scene-utils'
+import { mapDropNulls } from '../../../../../core/shared/array-utils'
+import * as EP from '../../../../../core/shared/element-path'
+import {
+  ElementInstanceMetadata,
+  ElementInstanceMetadataMap,
+} from '../../../../../core/shared/element-template'
+import {
+  CanvasPoint,
+  CanvasRectangle,
+  rectContainsPoint,
+  rectContainsPointInclusive,
+  Size,
+  size,
+  sizeFitsInTarget,
+} from '../../../../../core/shared/math-utils'
+import { ElementPath } from '../../../../../core/shared/project-file-types'
+import { AllElementProps } from '../../../../editor/store/editor-state'
+import { getAllTargetsAtPointAABB } from '../../../dom-lookup'
+import { InteractionCanvasState } from '../../canvas-strategy-types'
+import { AllowSmallerParent } from '../../interaction-state'
+import {
+  SingleAxisAutolayoutContainerDirections,
+  singleAxisAutoLayoutContainerDirections,
+} from '../flow-reorder-helpers'
+import { ReparentStrategy, ReparentSubjects, ReparentTarget } from './reparent-strategy-helpers'
+import { drawTargetRectanglesForChildrenOfElement } from './reparent-strategy-sibling-position-helpers'
+
+export type FindReparentStrategyResult = {
+  strategy: ReparentStrategy
+  isFallback: boolean
+  target: ReparentTarget
+}
+
+export function getReparentTargetUnified(
+  reparentSubjects: ReparentSubjects,
+  pointOnCanvas: CanvasPoint,
+  cmdPressed: boolean, // TODO: this should be removed from here and replaced by meaningful flag(s) (similar to allowSmallerParent)
+  canvasState: InteractionCanvasState,
+  metadata: ElementInstanceMetadataMap,
+  allElementProps: AllElementProps,
+  allowSmallerParent: AllowSmallerParent,
+): ReparentTarget | null {
+  const canvasScale = canvasState.scale
+
+  const validTargetparentsUnderPoint = findValidTargetsUnderPoint(
+    reparentSubjects,
+    pointOnCanvas,
+    cmdPressed,
+    canvasState,
+    metadata,
+    allElementProps,
+    allowSmallerParent,
+  )
+
+  // For Flex parents, we want to be able to insert between two children that don't have a gap between them.
+  const targetParentWithPaddedInsertionZone: ReparentTarget | null =
+    findParentByPaddedInsertionZone(
+      metadata,
+      validTargetparentsUnderPoint,
+      canvasScale,
+      pointOnCanvas,
+    )
+
+  if (targetParentWithPaddedInsertionZone != null) {
+    return targetParentWithPaddedInsertionZone
+  }
+
+  // fall back to trying to find an absolute element, or the "background" area of an autolayout container
+  const targetParentPath = validTargetparentsUnderPoint[0]
+  if (targetParentPath == null) {
+    // none of the targets were under the mouse, fallback return
+    return null
+  }
+
+  const targetParentUnderPoint: ReparentTarget = findParentUnderPointByArea(
+    targetParentPath,
+    metadata,
+    canvasScale,
+    pointOnCanvas,
+  )
+  return targetParentUnderPoint
+}
+
+function findValidTargetsUnderPoint(
```

Comment:
probably the most interesting change of the PR, this function was an inline filter function that was incredibly dense and mixed a few concepts and abstraction levels

Reply:
for comparison, the original inline array filtering can be found here https://github.com/concrete-utopia/utopia/blob/0452ce132ca5a1524e8685403d6922297dcdff8d/editor/src/components/canvas/canvas-strategies/strategies/reparent-strategy-helpers.ts#L320-L369